### ISSUE TITLE
Feature Request - Caching

### ISSUE THREAD

**Issue by rrichardson** [containous/traefik#878]
It would be nice to have caching without having to add additional moving parts to our infrastructure. Some features I'd like to see :  Cache by annotation/self-config.  LRU with max memory/item limit.  Max duration. 

It seems like it would be fairly simple to add a simple LRU caching middleware to Traefik.  The tricky part would be the configuration side, being able to specify which paths, duration, max mem, etc..  

Has this been discussed?

**Comment by errm** [comment]
Perhaps we could use https://github.com/lox/httpcache

**Comment by pascalandy** [comment]
Love this, but it might be a pain to maintain. 

IMHO, we could use existing caching solution like Varnish and Nginx. I don't know how to configure this at this point (Docker Swarm Mode here). We just need to know how to run this.

Any docker-compose setup anyone ?

**Comment by Berndinox** [comment]
+1

**Comment by voroninman** [comment]
> It seems like it would be fairly simple to add a simple LRU caching middleware to Traefik. The tricky part would be the configuration side, being able to specify which paths, duration, max mem, etc..

@rrichardson As a start memory limit should be enough to configure such middleware. A HTTP backend can control caching via response headers. The thing could go wild when you need to adjust hashing, purging, cookie.

**Comment by indigoid** [comment]
:+1: to simple caching that just has a single memory limit setting and honours `Cache-Control`/`Expires`. Any more than that would be too complicated and probably better suited to something like Varnish

**Comment by pascalandy** [comment]
Via a docker stack, it would be nice the have Varnish between the webapp and Traefik.

**Comment by jebw** [comment]
My use case would be traefic with let's encrypt and multiple dockerised web
apps behind it and traefic generating the config automatically.

I'd like to cache recently used static assets like I can in nginx. I can't
put varnish behind because it would require something to route requests to
the right apps, I can't put varnish in front because everything from
traefic forward is Https.

**Comment by fabiorauber** [comment]
Having cache we could have a Grace Period, in which Traefik would serve static pages even if the backend is down. This would eliminate the need for having Nginx/Varnish in front, at least in my scenario.

**Comment by funkyfuture** [comment]
just some 2 cents, that would be relevant for my use-cases and may have already been mentioned in some way in a previous post.

- i'd like to 'microcache' the responses of backends for a short period of time in order to relieve them (mainly when the application itself doesn't cache), e.g.
  - a response to a GET request that resolves as 200 is cached for 5 seconds
  - a response that returns in a 301 is cached for 10 minutes

- when appropriate (again, mainly 200 responses to GET requests), a cached result is returned to the client when the backends are timing out or return a 50x

- as i depend on traefik's acme client and tls handling, i cannot use nginx for such caching in front of traefik
- while putting several nginx instances between traefik and backends implies a bigger scaling footprint;  also configuration becomes tedious

**Comment by jonashackt** [comment]
I would also really really appreciate, if Traefik  would be able to cache! Please implement this :) For the meantime, if you look for a simple but fully working example how to use Nginx as a cache, have a look at https://github.com/jonashackt/traefik-cache-nginx-spring-boot

**Comment by joeldeteves** [comment]
This is a much-needed feature for production environments.

**Comment by semtex41** [comment]
I would also like to +1 this feature request. 
A single pane for the load balance + caching, that is easy to configure and docker/kubernetes friendly would be a game changer.

**Comment by wdalmut** [comment]
+1

**Comment by anton-johansson** [comment]
I agree with @joeldeteves. I'm currently testing Traefik without anything in front, and preparing for production use, but the cache part is important. With nginx in the front, we can easily handle cache of static content right there.

**Comment by pascalandy** [comment]
Are we sure Traefik is the right place to do this? Varnish and Nginx do cache out of the box. 

Should we instead work on a great `docker-compose` template that makes those two elements work together in a transparent way?

**Comment by anton-johansson** [comment]
I'm definitely not sure about it, but it would be great with a simple way to activate caching directly in the load balancer, just like nginx.

I'm not sure how it would work with a docker service in-between handling the cache, so I'd appreciate any examples or templates for that aswell. As long as it's easy to use I'm happy. :)

**Comment by joeldeteves** [comment]
@pascalandy IMO the power of traffic comes from labels. Being able to define caching behavior based on labels would be extremely powerful.

**Comment by pascalandy** [comment]
I have no idea as well. 

> I'm not sure how it would work with a docker service in-between handling the cache, so I'd appreciate any examples or templates for that aswell.

**Comment by isontheline** [comment]
This feature request could be again a killer feature for Træfik!

**Comment by Raboo** [comment]
Seriously, enough. This issue is open since 2016. There are a lot of subscribers like me that have notifications turned on. And most comments say the same thing. It is understood that this would be a killer feature, awesome feature, that YOU also want this, etc etc. So I suggest unless you have a PR or something constructive to bring to this discussion, just use the thumbs up on the original post and push the subscribe button to get an update of when this might be closed.

**Comment by lfniederauer** [comment]
Some thoughts:
Map the files of /var/www/webroot to a volume, they communicate over the swarm overlay.
Is really necessary the attachable on the networks?
Learning traefik, I had some issues with the cipher suite and it´s ordering. I can point to something but can´t help. The cipher  TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305" caught my attention.  I´ve ended up in this article: https://blog.gopheracademy.com/advent-2016/exposing-go-on-the-internet/
You still can bypass traefik for serving files and run side-by-side with a nginx reverse service.
You can use a redis service on the stack and use nginx cache.

**Comment by gp187** [comment]
Cool feature! Must be configurable by file type, origin, disable logging and passing other header rules
```
location ~ \.(js|css|png|jpg|jpeg|gif|ico|html|woff|woff2|ttf|svg|eot|otf)$ {
    add_header "Access-Control-Allow-Origin" "*";
    expires 1M;
    access_log off;
    add_header Cache-Control "public";
}```

**Comment by PureKrome** [comment]
Yep - originally requested back in 2017 and lots of chat in 2018. Still, no word from the Official Team 😢 

Would love _some_ acknowledgement  .. even if it is a "nope".

**Comment by pascalandy** [comment]
They decided it's a `priority/P3 on on Jun 6, 2017. You can disagree with that priority but you can't say they didn't say anything. They are highly responsive but they can't be highly responsive on every features around.

> still, no word from the Official Team. Would love some acknowledgement .. even if it is a "nope".

**Comment by PureKrome** [comment]
Sorry @pascalandy - i was meaning via some text/post comment-update. Not just a tag that feels like it's lost in the backlog.

I should have been more verbose with my comment, above. Sorry if that came off as rude :( soz.

**Comment by djmaze** [comment]
Don't know if this helps everyone, but I built a [configurable nginx proxy container](https://github.com/djmaze/docker-caching-proxy) which can easily be used in in conjunction with Traefik. Set the `UPSTREAM` to the app URL in your internal (overlay) network, then use the same `Host` rule as for your app service but with additional `PathPrefix` or `Path` rules.

**Comment by pySilver** [comment]
@PureKrome well its very complex (it's much more than @gp187 code snippet shows).  Giving that it's far from being core feature I'm not surprised it's still not there. Just use nginx, varnish or cdn (if possible)

**Comment by ronny35** [comment]
Hi,

Is it possible to implement caching fonctionnality ?
For example, like this https://github.com/victorspringer/http-cache

**Comment by notsureifkevin** [comment]
Greetings! There is a plugin for HTTP caching, while it's file-based, there are certainly opportunities for other implementations to be built which use fast data stores (Redis, Memcached, etc).

https://pilot.traefik.io/plugins/270947801855164928/simple-cache

**Comment by Darkweak** [comment]
Another plugin will be release soon with better integration based on my cache system called Souin
A minimal version is under development more efficient and as simple as possible to configure.
The full version handles redis, in memory and is easy to introduce your own cache provider since you implement the required interface.
Cf this repository: https://github.com/darkweak/souin

**Comment by Roken120** [comment]
Hello, any news for caching ?

**Comment by Darkweak** [comment]
I'm currently working on the RFC support and shared cache support too. Then I'm waiting for the support of unsafe lib to publish my cache system as plugin

**Comment by misaon** [comment]
We solved the caching of assets in Traefik for our purposes. We have all the images in "/media/cache" and all the CSS/JS in the "/assets/" folder. Then all you have to do is create more guests and define custom headers for one of them.

(Traefik sorts route priority by length of route "DESC")

My app labels:
```yaml
labels:
    - "traefik.enable=true"
    - "traefik.docker.network=traefik-public"
    # Static content
    - "traefik.http.routers.my-app-static.rule=Host('my-app.com') && PathPrefix(`/assets/`, `/media/cache/`)"
    - "traefik.http.routers.my-app-static.entrypoints=https"
    - "traefik.http.routers.my-app-static.tls=true"
    - "traefik.http.routers.my-app-static.middlewares=cache-headers"
    # Dynamic content
    - "traefik.http.routers.my-app.rule=Host('my-app.com')"
    - "traefik.http.routers.my-app.entrypoints=https"
    - "traefik.http.routers.my-app.tls=true"
    - "traefik.http.services.my-app.loadbalancer.server.port=80"
```

My Traefik service labels:
```yaml
labels:
    - traefik.http.middlewares.cache-headers.headers.customresponseheaders.Cache-Control=public,max-age=2592000
```

You can define your own hosts regex for check jpg, css, js etc.

Works great together with Traefik GZIP middleware! :rocket:

**Comment by Darkweak** [comment]
@misaon Your solution doesn't fix the Træfik problem that tries to fetch service instead of serving app cache. Then it's not RFC compliant due to missing TTL, Vary and other headers defined in the RFC

**Comment by misaon** [comment]
@Darkweak sure, this not solve issue for all peoples in this topic, but you can determine assets by route and set them needed headers. This example is for our purposes. We steel waiting for RFC implementation on Traefik side.

**Comment by Darkweak** [comment]
It'll be implemented as soon as Træfik supports go unsafe library for plugins then I'll submit my cache system as official plugin. ATM it works out of the box as standalone software.

**Comment by Darkweak** [comment]
[Souin](https://github.com/darkweak/souin) now supports [RFC-7234](https://tools.ietf.org/html/rfc7234) out of the box. I'm now waiting for the `unsafe` library from Træfik side to publish efficient cache system as plugin

**Comment by PureKrome** [comment]
@Darkweak wow - the Souin app looks great!

**Comment by JaneJeon** [comment]
Just my 2 cents, I’d love Varnish *integration* with Traefik via some middleware, but I really don’t want the Traefik team to reinvent the wheel and re-implement the caching logic/language/config within Traefik itself!

**Comment by osminogin** [comment]
A bit of a strange discussion. Developing the thousandth web server is okay and not reinventing the wheel. And the caching functionality in this web server is a wheel reinvention...

**Comment by Darkweak** [comment]
@JaneJeon it would be supported using Souin as Træfik plugin when they will fix this yaegi issue https://github.com/traefik/yaegi/issues/1172

**Comment by darkweak** [comment]
Hello there, little update about this, [Souin](https://github.com/darkweak/souin) is now available as [træfik plugin here](https://pilot.traefik.io/plugins/6123af58d00e6cd1260b290e/souin) 🎉. You can now use an efficient HTTP cache system without any changes on your stack.

**Comment by yongzhang** [comment]
Would like to see this in community edition since it's a really useful and important feature.

**Comment by thanhtoan1196** [comment]
any updates?

**Comment by tigerblue77** [comment]
[This is an enterprise feature](https://doc.traefik.io/traefik-enterprise/middlewares/http-cache/). It seems that you have to buy Traefik enterprise to get caching functionality, tell me if I'm wrong ?

**Comment by thanhtoan1196** [comment]
> [This is an enterprise feature](https://doc.traefik.io/traefik-enterprise/middlewares/http-cache/) so, unfortunately, it seems that you have to buy Traefik enterprise to get caching functionality. Don't hesitate to tell me if I'm wrong.

unfortunately :(

**Comment by nmengin** [comment]
Hey @thanhtoan1196, @tigerblue77,

This issue is tagged as `priority/P3` `contributor/wanted`, which means that even if it's not part of our roadmap for Traefik, we'll consider every contribution from the community on this topic.
